### PR TITLE
[13.0][IMP] product_supplierinfo_for_customer: add `_select_customerinfo

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
The idea of this method is to centralize the searches of customerinfo
to have more consistent usage in child modules.

@ForgeFlow

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
cc @pedrobaeza

File: product_supplierinfo_for_customer/models/product_product.py [hunk 1]
```diff
@@ -114,3 +103,30 @@ def price_compute(self, price_type, uom=False, currency=False, company=False):
...
+    ):
+        """Customer version of the standard `_select_seller`. """
+        # TODO: For now it is just the function name with same arguments, but
+        #  can be changed in future migrations to be more in line Odoo
+        #  standard way to select supplierinfo's.
+        if not params:
+            params = dict()
+        params.update({"partner_id": partner.id})
+        res = self._prepare_customerinfo(params)
+        return res[:1]
```
Comment (chienandalu) [inline product_supplierinfo_for_customer/models/product_product.py]: Maybe a limit=1 in the domain would suffice?

**Review by chienandalu — COMMENTED** [review]
Looks less risky indeed :sweat_smile:

File: product_supplierinfo_for_customer/models/product_product.py [hunk 2]
```diff
@@ -60,38 +60,27 @@ def _name_search(
         res.extend(self.browse(product_ids).name_get())
         return res
 
-    def _get_price_from_customerinfo(self, partner_id):
+    def _get_price_from_customerinfo(self, partner):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer/models/product_product.py]: It's still risky, as you are changing the signature of one method, and thus other extensions will fail in this. Please keep the same signature as previously. You can change the signature on a version migration.

File: product_supplierinfo_for_customer/models/product_product.py [hunk 3]
```diff
@@ -114,3 +103,30 @@ def price_compute(self, price_type, uom=False, currency=False, company=False):
         return super(ProductProduct, self).price_compute(
             price_type, uom, currency, company
         )
+
+    def _prepare_customerinfo(self, params):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer/models/product_product.py]: Why this double method? If you want to inject things on the domain for performing only one search, then put a `_prepare_domain` hook, but this is not adding anything useful to the inheritance (nothing that you can't done on the main method).

File: product_supplierinfo_for_customer/models/product_product.py (hunk 1 above)
Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py]: I want to keep the structure like [_sellect_sellers](https://github.com/odoo/odoo/blob/14.0/addons/product/models/product.py#L602) as much as possible

File: product_supplierinfo_for_customer/models/product_product.py (hunk 2 above)
Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py]: Done

File: product_supplierinfo_for_customer/models/product_product.py (hunk 3 above)
Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py]: Done.

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza does it look good now?

File: product_supplierinfo_for_customer/models/product_product.py (hunk 1 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py]: This is not adding anything useful, as the return result is going to be the same 1 record, but with less performance and readability. Not always the upstream code is the one to imitate...

File: product_supplierinfo_for_customer/models/product_product.py [hunk 4]
```diff
@@ -64,34 +64,23 @@ def _get_price_from_customerinfo(self, partner_id):
...
-                "|",
-                ("product_id", "=", self.id),
-                "&",
-                ("product_tmpl_id", "=", self.product_tmpl_id.id),
-                ("product_id", "=", False),
-            ],
-            limit=1,
-            order="product_id, sequence",
-        )
+        if partner_id and isinstance(partner_id, int):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer/models/product_product.py]: `partner_id` shouldn't be polymorphic and only receives an integer IMO.

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py]: can you make this concession at least? this won't affect in any way any installation...

Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py]: Yes, it does. Any subsequent inherited method will need to check the same each time.

File: product_supplierinfo_for_customer/models/product_product.py (hunk 1 above)
Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py]: Ok, done.

File: product_supplierinfo_for_customer/models/product_product.py (hunk 4 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py]: That's why type hinting was added in Python 3.x, for avoiding these ugly hacks.

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:67]: Ok, what should I put then?

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:67]:
```
...
partner = self.env["res.partner"].browse(partner_id)
customerinfo = self._select_customerinfo(partner=partner)
```

will this be ok?

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py]: pushed, hope it is good now

**Comment by LoisRForgeFlow** [comment]
@chienandalu could you check again?

File: product_supplierinfo_for_customer/models/product_product.py (hunk 4 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py]: As the argument is the integer, always pass the ID, and do the browse if you need the recordset. On a migration to next non migrated version, the argument may be changed to `partner`.

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py]: it is done, could you check?

File: product_supplierinfo_for_customer/models/product_product.py [hunk 5]
```diff
@@ -64,34 +64,22 @@ def _get_price_from_customerinfo(self, partner_id):
...
         if customerinfo:
             return customerinfo.price
         return 0.0
 
     def price_compute(self, price_type, uom=False, currency=False, company=False):
         if price_type == "partner":
-            partner_id = self.env.context.get(
-                "partner_id", False
-            ) or self.env.context.get("partner", False)
-            if partner_id and isinstance(partner_id, models.BaseModel):
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer/models/product_product.py]: This polymorphism seems to be still needed (what I was talking about is to not create new ones).

File: product_supplierinfo_for_customer/models/product_product.py [hunk 6]
```diff
@@ -64,34 +64,22 @@ def _get_price_from_customerinfo(self, partner_id):
...
-                partner_id = partner_id.id
+            partner = self.env.context.get("partner_id", False) or self.env.context.get(
+                "partner", False
+            )
```
Comment (MiquelRForgeFlow) [inline product_supplierinfo_for_customer/models/product_product.py:77]: you shouldn't touch this. `_get_price_from_customerinfo` receives int, not a record

Comment (MiquelRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:77]: Oh, pedro was faster haha

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza Can you check again?

File: product_supplierinfo_for_customer/models/product_product.py [hunk 7]
```diff
@@ -114,3 +104,30 @@ def price_compute(self, price_type, uom=False, currency=False, company=False):
         return super(ProductProduct, self).price_compute(
             price_type, uom, currency, company
         )
+
+    def _prepare_domain_customerinfo(self, params):
+        partner_id = params.get("partner_id")
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer/models/product_product.py:110]:
```suggestion
        self.ensure_one()
        partner_id = params.get("partner_id")
```

**Review by pedrobaeza — APPROVED** [review]

File: product_supplierinfo_for_customer/models/product_product.py (hunk 7 above)
Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:110]: added.

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza Good to merge?

**Review by chienandalu — APPROVED** [review]

**Comment by pedrobaeza** [comment]
/ocabot merge patch

I think we are in time for changing params in v14, as it has been just merged.

**Pull request merged**